zebra-state/src/service/finalized_state.rs

```diff
@@ -259,10 +265,11 @@ impl FinalizedState {
         // state, these will contain the empty trees.
         let mut sapling_note_commitment_tree = self.sapling_note_commitment_tree();
         let mut orchard_note_commitment_tree = self.orchard_note_commitment_tree();
+        let mut history_tree = self.history_tree();
 
         // We use a closure so we can use an early return for control flow in
         // the genesis case
-        let prepare_commit = || -> rocksdb::WriteBatch {
+        let prepare_commit = || -> Result<rocksdb::WriteBatch, BoxError> {
```

Comment:
Missing docs:

Can you add a comment explaining how errors cancel the batch?
I'm pretty sure it works, but it took me a while to think through it.

It might also be helpful to add error and panic docs to the method documentation.

Reply:
I decided to revert to not returning an error in 949a5e5c. We were already `expect`ing the note commitment trees and there was no reason to not do the same for the history tree.

Reply:
I think we need to revert this change, but the reasoning is pretty complex. So I'd like you to double-check.

We'll also need to add comments documenting this information.

`commit_finalized_direct` handles both checkpointed and non-finalized blocks. (Needs docs.)

We can `expect` on the note commitment trees, because notes are bound by the non-malleable v5 transaction ID. (Needs docs.) Transaction IDs are bound to the transaction merkle root header field, block hash, chain of previous block hashes, and checkpoint list by the checkpoint verifier.

But we can't `expect` on the chain history hash, because the height data that's used in the chain history is only bound by the authorizing data commitment. (Needs docs.) Coinbase heights are stored as bitcoin scripts.
https://zips.z.cash/zip-0244#a-1-transparent-scripts-digest

There's an alternate height field in v5 transactions that's covered by the transaction ID. But Zebra doesn't check it yet (#2387), or use it for our block heights.

When we check the block commitment, we'll also need to be able to return an error here. Because it binds the authorizing data commitment, which binds the transaction authorizing data in the block. And if it's a checkpoint block, we haven't checked that data yet.

See also my detailed answer here, starting with:
> Unfortunately not, because we need to make sure that v5 transactions in checkpointed blocks haven't been modified

https://github.com/ZcashFoundation/zebra/issues/2134#issuecomment-888709631